### PR TITLE
feat:[HL-69] Saved Recipes <> SavedRecipesViewController & SavedRecipesViewModel 

### PR DESCRIPTION
**Pull request by esraakhaled**
#### Proposed Changes

1.create SavedRecipeModel
2.create SavedRecipesController using UITableViewDiffableDataSource
3.Create SavedRecipeViewModel
4.Add ActualCell For applying diffable

#### Screenshot
Required for any UI changes

Simulator Screen Shot - iPhone 14 Pro - 2023-07-06 at 03 12 06

#### Checklist

- [x] I have tested these changes locally.
- [ ] I have added appropriate documentation or updated existing documentation.
- [x] I have added appropriate test coverage or updated existing test coverage.
- [ ] I have updated the changelog (if applicable).
- [x] I have followed the project's code style and formatting guidelines.
- [ ] I have reviewed and adhered to the project's contributing guidelines.

### CODE REVIEW

File: Healthy/Classes/Entities/SavedRecipe.swift [hunk 1]
```diff
@@ -1 +1,46 @@
-struct SavedRecipe {}
+import UIKit
+
+struct SavedRecipe: Hashable, Equatable {
+
+    // MARK: - Properties
+
+    let id = UUID()
+    /// The title of the recipe
```
Comment (ahmdmhasn) [inline Healthy/Classes/Entities/SavedRecipe.swift]:
```suggestion
    let id = UUID()

    /// The title of the recipe
```

File: Healthy/Classes/Entities/SavedRecipe.swift [hunk 2]
```diff
@@ -1 +1,46 @@
...
+
+init(title: String?, recipeImage: UIImage?, rating: Double?,
+     chefName: String?, cookingTime: Int?, toggleBookmark: @escaping () -> Void) {
+        self.title = title
+        self.recipeImage = recipeImage
+        self.rating = rating
+        self.chefName = chefName
+        self.cookingTime = cookingTime
+        self.toggleBookmark = toggleBookmark
+    }
```
Comment (ahmdmhasn) [inline Healthy/Classes/Entities/SavedRecipe.swift]:
```suggestion
    }
```

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift [hunk 3]
```diff
@@ -36,15 +36,33 @@ final class SavedRecipesTableViewCell: UITableViewCell {
         configureLayout()
     }
 
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        contentView.frame = contentView.frame.inset(by: UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10))
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift:29]: Let's add this to `awakeFromNib`

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift [hunk 4]
```diff
@@ -36,15 +36,33 @@ final class SavedRecipesTableViewCell: UITableViewCell {
...
+        if let chefName = viewModel.chefName {
+            chefNameLabel.text = chefName
+        }
+
+        if let cookingTime = viewModel.cookingTime {
+            timeLabel.text = "\(cookingTime) min"
+        }
         // TODO: Set rate count.
-        toggleBookmark = viewModel.toggleBookmark
+         toggleBookmark = viewModel.toggleBookmark
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift]:
```suggestion
        toggleBookmark = viewModel.toggleBookmark
```

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift [hunk 5]
```diff
@@ -36,15 +36,33 @@ final class SavedRecipesTableViewCell: UITableViewCell {
...
+            titleLabel.text = title
+        }
+
+        if let chefName = viewModel.chefName {
+            chefNameLabel.text = chefName
+        }
+
+        if let cookingTime = viewModel.cookingTime {
+            timeLabel.text = "\(cookingTime) min"
+        }
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift]:
```suggestion
        recipeImageView.image = viewModel.recipeImage
        titleLabel.text = viewModel.title
        chefNameLabel.text = viewModel.chefName
        
        if let cookingTime = viewModel.cookingTime {
            timeLabel.text = "\(cookingTime) min"
        } else {
            timeLabel.text = nil        
        }
```

Text and image are optional. Optional chaining.

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift [hunk 6]
```diff
@@ -36,15 +36,33 @@ final class SavedRecipesTableViewCell: UITableViewCell {
...
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        contentView.frame = contentView.frame.inset(by: UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10))
+    }
+
     // MARK: Configure
 
-    func update(with viewModel: ViewModel) {
+    func update(with viewModel: SavedRecipe) {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift:34]: Please revert this change. 

Now the cell is depend on the `SavedRecipe` which is not correct!

File: Healthy/Classes/Entities/SavedRecipe.swift [hunk 7]
```diff
@@ -1 +1,46 @@
...
+    /// The rating of the recipe
+    let rating: Double?
+
+    /// The chef  who posted the recipe
+    let chefName: String?
+
+    /// The cooking Time of The recipe
+    let cookingTime: Int?
+
+    var toggleBookmark: () -> Void
```
Comment (ahmdmhasn) [inline Healthy/Classes/Entities/SavedRecipe.swift:24]:
```suggestion
```

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift [hunk 8]
```diff
@@ -1,12 +1,22 @@
...
 
 final class SavedRecipesViewController: UIViewController {
 
+    enum Section {
+        case main
+    }
+
     // MARK: Outlets
 
+    @IBOutlet weak var navigationBar: UINavigationBar!
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:12]:
```suggestion
```

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift [hunk 9]
```diff
@@ -33,7 +48,42 @@ extension SavedRecipesViewController {}
 
 // MARK: - Configurations
 
-extension SavedRecipesViewController {}
+extension SavedRecipesViewController {
+    func configureNavigationBar() {
+        navigationBar.topItem?.title = "Saved recipes"
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:53]:
```suggestion
        title = "Saved recipes"
```

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift [hunk 10]
```diff
@@ -33,7 +48,42 @@ extension SavedRecipesViewController {}
...
+            UINavigationBar.appearance().titleTextAttributes = [NSAttributedString.Key.font: font]
+        }
+    }
+
+    func configureTableView() {
+       tableView.registerNib(cell: SavedRecipesTableViewCell.self)
+    }
+
+    func configureTableViewDataSource() {
+    dataSource = UITableViewDiffableDataSource<Section, SavedRecipe>(tableView: tableView) { tableView, _, viewModel in
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift]: Indentation...Indentation...Indentation...Indentation

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift [hunk 11]
```diff
@@ -33,7 +48,42 @@ extension SavedRecipesViewController {}
...
+        let cell = tableView.dequeueReusableCell(withIdentifier: SavedRecipesTableViewCell.reuseIdentifier)
+            as? SavedRecipesTableViewCell
+            cell?.update(with: viewModel)
+            return cell
+        }
+    }
+
+    func updateTableData(viewModel: [SavedRecipe]) {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:70]:
```suggestion
    func updateTableData(viewModel: [SavedRecipesTableViewCell.ViewModel]) {
```

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift [hunk 12]
```diff
@@ -33,7 +48,42 @@ extension SavedRecipesViewController {}
 
 // MARK: - Configurations
 
-extension SavedRecipesViewController {}
+extension SavedRecipesViewController {
+    func configureNavigationBar() {
+        navigationBar.topItem?.title = "Saved recipes"
+        if let font = UIFont(name: "Poppins-Bold", size: 18) {
+            UINavigationBar.appearance().titleTextAttributes = [NSAttributedString.Key.font: font]
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift]: This will apply this font to the whole app!

Instead, please add this to Extensions/UINavigationBar+Appearance

extension UINavigationBar {
    
    /// Apply the default navigation bar appearance
    ///
    class func applyDefaultAppearance() {
        UINavigationBar.appearance().titleTextAttributes = [.font: UIFont.mediumRegular]
    }
}

and call it in the app delegate - did finish launching with options

File: HealthyTests/Classes/Modules/SavedRecipes/SavedRecipesViewModelTests.swift [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    // MARK: Properties
+    var viewModel: SavedRecipesViewModel!
+    var subscriptions = Set<AnyCancellable>()
+
+    override func setUpWithError() throws {
+        viewModel = SavedRecipesViewModel()
+    }
+
+    override func tearDownWithError() throws {
+        subscriptions.removeAll()
```
Comment (ahmdmhasn) [inline HealthyTests/Classes/Modules/SavedRecipes/SavedRecipesViewModelTests.swift:16]: 🚀🚀

**Review by ahmdmhasn — APPROVED** [review]
Few comments to be done. GEART PROGRESS! 🎉🚀

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift (hunk 11 above)
Comment (esraakhaled) [reply Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:70]: Model SavedRecipe Conforms to Hashable & Equatable and the
 static func == (lhs: SavedRecipe, rhs: SavedRecipe) -> Bool {
        return  lhs.title == rhs.title &&
        lhs.rating == rhs.rating &&
        lhs.chefName == rhs.chefName &&
        lhs.cookingTime == rhs.cookingTime &&
        lhs.recipeImage == rhs.recipeImage
    }
is required in unit testing to test assertEqual of obj
all this is already on SavedRecipe Model

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift (hunk 3 above)
Comment (esraakhaled) [reply Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift:29]: Simulator Screen Shot - iPhone 14 Pro - 2023-07-06 at 02 47 33
if putting it in awakeFromNib instead of layoutSubviews

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift (hunk 9 above)
Comment (esraakhaled) [reply Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:53]: Simulator Screen Shot - iPhone 14 Pro - 2023-07-06 at 02 47 33
changes title of maintabbar not title of nib file

**Review by abdallah-7698 — APPROVED** [review]

File: Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift (hunk 11 above)
Comment (ahmdmhasn) [reply Healthy/Classes/Modules/SavedRecipes/SavedRecipesViewController.swift:70]: This has not been applied, please recreate another pull request and address this comment.

@esraakhaled

File: Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift [hunk 14]
```diff
@@ -1,18 +1,5 @@
...
-    struct ViewModel {
-        var recipeImageURL: URL
-        var title: String
-        var chefName: String
-        var rate: Double
-        var time: String
-        var toggleBookmark: () -> Void
-    }
-}
-
```
Comment (ahmdmhasn) [inline Healthy/Classes/Modules/SavedRecipes/Cell/SavedRecipesTableViewCell.swift:15]: Should have never be removed, please bring it back in a follow pr.

File: Healthy/Classes/Entities/SavedRecipe.swift (hunk 7 above)
Comment (ahmdmhasn) [reply Healthy/Classes/Entities/SavedRecipe.swift:24]: No addressed! @esraakhaled
